framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java

```diff
@@ -3442,4 +3547,32 @@ public ProcessingEnvironment getProcessingEnv() {
     public CFContext getContext() {
         return checker;
     }
+
+    /**
+     * Factory method to reflectively set {@link AnnotatedTypeFactory#viewpointAdaptor}. By default,
+     * this method gets the name of type system from the canonical name of "this", and concatenate
+     * it with "ViewpointAdaptor". If the concrete type system has a different naming mechanism,
+     * then developer of that type system should override this method to instantiate it correctly.
+     * If the target class can be instantiated successfully, it's used. Otherwise, internal error is
+     * triggered and will be error aborted.
+     *
+     * @return ViewpointAdaptor newly created
+     */
+    protected ViewpointAdaptor<?> createViewpointAdaptor(
+            boolean claimsToEnableViewpointAdaptation) {
+        if (!claimsToEnableViewpointAdaptation) return null;
+        // Reflectively lookup subclass of {@link ViewpointAdaptor} using type factory naming convention
+        String clazzName =
+                this.getClass()
+                        .getCanonicalName()
+                        .replace("AnnotatedTypeFactory", "ViewpointAdaptor");
+        try {
```

Comment:
We should see whether we can move BaseTypeChecker#invokeConstructorFor to a better location and reuse similar logic here. Can you see what logic we use when creating the Visitor, AnnotatedTypeFactory, and other classes similar to this one?